source/common/config/config_provider_impl.h

```diff
@@ -0,0 +1,364 @@
+
+private:
+  SystemTime last_updated_;
+  ConfigProviderManagerImpl& config_provider_manager_;
+};
+
+class DynamicConfigProviderImpl;
+
+/**
+ * Provides generic functionality required by all dynamic ConfigProvider subscriptions, including
+ * shared lifetime management via shared_ptr.
+ *
+ * To do so, this class keeps track of a set of dynamic config providers associated with an
+ * underlying subscription; providers are bound/unbound as needed as they are created and destroyed.
+ * This is done to avoid exposing shared ownership semantics to the external API (see
+ * ConfigProvider).
+ *
+ * This class can not be instantiated directly; instead, it provides the foundation for
+ * dynamic config subscription implementations which derive from it.
+ */
+class ConfigSubscriptionInstance : public Init::Target,
+                                   protected Logger::Loggable<Logger::Id::config> {
+public:
+  struct LastConfigInfo {
+    uint64_t last_config_hash_;
+    std::string last_config_version_;
+  };
+
+  ConfigSubscriptionInstance() = delete;
+
+  virtual ~ConfigSubscriptionInstance();
+
+  // Init::Target
+  void initialize(std::function<void()> callback) override {
+    initialize_callback_ = callback;
+    start();
+  }
+
+  /**
+   * Starts the subscription corresponding to a config source.
+   * A derived class must own the configuration proto specific Envoy::Config::Subscription to be
+   * started.
+   */
+  virtual void start() PURE;
+
+  const SystemTime& lastUpdated() const { return last_updated_; }
+
+  const absl::optional<LastConfigInfo>& configInfo() const { return config_info_; }
+
+  /**
+   * Must be called by derived classes when the onConfigUpdate() callback associated with the
+   * underlying subscription is issued.
+   */
+  void onConfigUpdate() {
+    setLastUpdated();
+    runInitializeCallbackIfAny();
+  }
+
+  /**
+   * Must be called by derived classes when the onConfigUpdateFailed() callback associated with the
+   * underlying subscription is issued.
+   */
+  void onConfigUpdateFailed() { runInitializeCallbackIfAny(); }
+
+  /**
+   * Determines whether a configuration proto is a new update, and if so, propagates it to all
+   * config providers associated with this subscription.
+   * @param config_proto supplies the newly received config proto.
+   * @param config_name supplies the name associated with the config.
+   * @param version_info supplies the version associated with the config.
+   * @return bool false when the config proto has no delta from the previous config, true otherwise.
+   */
+  bool checkAndApplyConfig(const Protobuf::Message& config_proto, const std::string& config_name,
+                           const std::string& version_info);
+
+  const std::unordered_set<const DynamicConfigProviderImpl*> dynamic_config_providers() const {
+    return std::unordered_set<const DynamicConfigProviderImpl*>(dynamic_config_providers_.begin(),
```

Comment:
Why do you need copy here? Seems you're only checking the size and the first element when you use it.

Reply:
Removed this accessor and replaced it getAnyBoundDynamicConfigProvider().